src/adaptors/multi_product.rs

```diff
@@ -136,9 +139,12 @@ where
 
     fn count(self) -> usize {
         match self.0 {
-            None => 0,
+            None => 0, // The cartesian product has ended.
```

Comment:
Would it make sense to have an `enum Something { MultiProductInner(iters, cur), CartesianProductHasEnded }` instead of an `Option` here?

Reply:
It makes sense and the other too. But it's just an option with customized names and without the methods and `?` shortcut.
On "permutations", I remember doing the opposite change and really simplify things with it.
In `next`, there is `let inner = self.0.as_mut()?;` that would be quite complicated with such change.

Reply:
Honestly I hesitated to add a comment here as the `None` variant for `.0` is already documented.

Reply:
A compromise:
```rust
pub use {
    Option as State,
    Option::None as ProductEnded,
    Option::Some as ProductInProgress,
};
```